File: weekly-contest-159/3-replace-the-substring-for-balanced-string/sliding-window.js
/**
 * 滑动窗口：窗口用以减小字符频次，求使剩余字符串频次合法的最小窗口
 * 
 * 时间：O(N), 348ms
 */
var balancedString = function (s) {
  const count = new Map()
  for (const ch of s) {
    count.set(ch, (count.get(ch) || 0) + 1)
  }
  
  let limit = s.length / 4 // 每种字符最多出现limit次
  let [left, right] = [0, 0]
  let res = s.length

  while (right < s.length) {
    const deleteChar = s[right++]
    count.set(deleteChar, count.get(deleteChar) - 1)

    while (left <= right && isValid(count, limit)) { // 合法，则尝试缩小窗口
      res = Math.min(res, right - left)
      const releaseChar = s[left++]
      count.set(releaseChar, (count.get(releaseChar) || 0) + 1)
    }
  }
  
  return res
};

function isValid (count, limit) {
  return Array.from(count.values()).every(cnt => cnt <= limit)
}

[
  'QWER',
  'QQWE',
  'QQQW',
  'QQQQ',
].forEach(s => {
  console.log(balancedString(s))
})

File: weekly-contest/weekly-contest-159/3-replace-the-substring-for-balanced-string/sliding-window.cpp
#include <string>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    int balancedString(string s) {
        n = s.size();
        freq.clear();

        for (auto ch : s) {
            ++freq[ch];
        }
        if (check()) return 0; // 本来就平衡了

        int j = 0;
        int res = n;
        
        for (int i = 0; i < n; ++i) {
            --freq[s[i]];

            // 如果剩下的平衡，则可以收敛
            while (check()) {
                res = min(res, i - j + 1); // 更新答案
                ++freq[s[j]];
                ++j;
            }
        }

        return res;
    }

    // 检查（除去窗口的）剩余字符串是否平衡
    bool check () {
        int limit = n / 4;
        for (char ch : "QWER") {
            if (freq[ch] > limit) return false;
        }
        return true;
    }

private:
    int n;
    unordered_map<char, int> freq; // 字母 => 个数
};